server/lib/dataone/fakeoauth.py

```diff
@@ -0,0 +1,46 @@
+from urllib.parse import urlparse, urlunparse, quote
+from d1_common.env import D1_ENV_DICT
+
+from girder.api.rest import getApiUrl
+from girder.plugins.oauth.providers.base import ProviderBase
+
+# Because, why would they keep up to date list of CNs and deployments?
+D1_ENV_DICT["stage2"] = dict(base_url="https://cn-stage-2.test.dataone.org/cn")
+
+
+class FakeDataONE(ProviderBase):
+    @classmethod
+    def get_cn(cls):
+        key = cls.__name__.lower().replace("dataone", "")
+        return D1_ENV_DICT[key]["base_url"]
+
+    @classmethod
+    def getUrl(cls, state):
+        _, _, redirect = state.partition(".")
+
+        url = "/".join(
+            (getApiUrl(), "account", cls.getProviderName(external=False), "callback")
+        )
+        url += "?state={}&code=dataone".format(quote(state))
+        auth_url = urlparse(cls.get_cn())._replace(
+            path="/portal/oauth", query="action=start&target={}".format(quote(url))
+        )
+        return urlunparse(auth_url)
+
+    def getToken(self, code):
+        return {
+            "provider": self.getProviderName(external=False),
+            "resource_server": urlparse(self.get_cn()).netloc,
+            "token_type": "dataone-pre",
+            "access_token": "",
+        }
+
+    def getClientIdSetting(self):
+        return "fake_client_id"
+
+    def getClientSecretSetting(self):
+        return "fake_secret_id"
+
+    @classmethod
+    def getTokenUrl(cls):
+        return urlunparse(urlparse(cls.get_cn())._replace(path="/portal/token"))
```

Comment:
If this `._replace` is replacing the full path suffix, then the full path should (correctly) be `https://cn-stage-2.test.dataone.org/portal/token`

Somehow this is still yielding a 404 for Girder :/

Reply:
Scratch that, reading the error in the token shows a mangled URL suffix and hints at the problem: `&#47;portal&#47;token&amp;target=https%3A&#47;&#47;girder.local.wholetale.org&#47;api&#47;v1&#47;account&#47;dataonestage2&#47;callback%3Fstate%3D9t8QNP1PCO0HqtBv8uGOhyXGfQNtj1haqFuLr1Iit7Vb5XGoIck0DivBHJVAzLpx.https%253A&#47;&#47;girder.local.wholetale.org&#47;%2523ext_keys%26code%3Ddataone`

This long URL suffix should just be `/portal/token`, as there does not appear to be a `target` parameter on this page (NOTE: this is not a REST API, this is just Apache Tomcat)

Navigating to https://cn-stage-2.test.dataone.org/portal/token&target=https%3A%2F%2Fgirder.local.wholetale.org%2Fapi%2Fv1%2Faccount%2Fdataonestage2%2Fcallback%3Fstate%3D9t8QNP1PCO0HqtBv8uGOhyXGfQNtj1haqFuLr1Iit7Vb5XGoIck0DivBHJVAzLpx.https%253A%2F%2Fgirder.local.wholetale.org%2F%23ext_keys%26code%3Ddataone in the browser yields the same 404, but dropping the query string altogether yields a 200 response

Reply:
`/portal/token` endpoint should be called without any parameters. I'm confused what's going on. I'm gonna retest this branch to see if something's changed.

Reply:
@bodom0015 I tested it just now. Requests you should be seeing:

1. After clicking button for unauthorized state it goes to:

```
https://cn-stage-2.test.dataone.org/portal/oauth?action=start&target=https%3A//girder.local.wholetale.org/api/v1/account/dataonestage2/callback%3Fstate%3DkfL3zSIOrMPyWi0H.https%253A//girder.local.wholetale.org/%2523ext_keys%26code%3Ddataone
```

2. This goes back to `#ext_keys` page and in the dev console you can see a list of providers being outputed:

```
(4) [{…}, {…}, {…}, {…}]
0: {docs_href: "https://zenodo.org/account/settings/applications/tokens/new/", fullName: "Zenodo", logo: "/9j/4AAQSkZJRgABAQAAAQABAAD/2wBDAAMCAgICAgMCAgIDAw…AQBAEAQBAEAQBAEAQBAEAQBAEAQBAEAQBAEAQBAEAQBAEB//Z", name: "zenodo", tags: Array(2), …}
1: {fullName: "ORCID", logo: "/9j/4AAQSkZJRgABAQAAAQABAAD/2wCEAAkGBxAPDw8PEA8PDh…5ciIeUjyXJQitFXtE9VV3oj20EoulKIBCBJIAAAAAAAAAP//Z", name: "orcid", state: "unauthorized", tags: Array(1), …}
2:
fullName: "DataONE Stage 2 CN"
logo: "/9j/4AAQSkZJRgABAQAAAQABAAD/4gKgSUNDX1BST0ZJTEUAAQ"
name: "dataonestage2"
state: "preauthorized"
tags: ["publish"]
type: "dataone"
url: "https://cn-stage-2.test.dataone.org/portal/token"
__proto__: Object
3: {docs_href: "http://guides.dataverse.org/en/latest/api/auth.html", fullName: "Dataverse", logo: "/9j/4AAQSkZJRgABAQEASABIAAD/4QCARXhpZgAASUkqAAgAAA…qkXWwfFPrtoD2lTXkJhB6kKgyYLNtZIelA8KBgNgID/Ov/9k=", name: "dataverse", tags: Array(2), …}
length: 4
```
Note: state "preauthorized".
3. New state should trigger additional call to `/portal/token` for state "preauthorized":

```
https://cn-stage-2.test.dataone.org/portal/token
```
This gives 200 for me.

4. Token should be printed in the console :

```
Response from DataONE:
eyJhbGciOiJ...NEoTo866w
```